src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs

```diff
@@ -0,0 +1,398 @@
+                                ``{`` [] ``}`` :> MemberDeclarationSyntax
+            let implMany = ``method`` "IQArray<Qubit>" name ``<<`` [] ``>>`` 
+                                ``(`` [``param`` "n" ``of`` intType] ``)`` [``public``;``abstract``]
+                                ``{`` [] ``}`` :> MemberDeclarationSyntax
+            seq { implOne; implMany }
+        let buildReleaseMethods name =
+            let implOne = ``method`` "void" name ``<<`` [] ``>>`` 
+                                ``(`` [``param`` "qb" ``of`` qubitType] ``)`` [``public``;``abstract``]
+                                ``{`` [] ``}`` :> MemberDeclarationSyntax
+            let implMany = ``method`` "void" name ``<<`` [] ``>>`` 
+                                ``(`` [``param`` "qbs" ``of`` qubitArrayType] ``)`` [``public``;``abstract``]
+                                ``{`` [] ``}`` :> MemberDeclarationSyntax
+            seq { implOne; implMany }
+        let dumpMachineMethod = ``method`` "void" "DoDumpMachine" ``<<`` ["T"] ``>>``
+                                    ``(`` [``param`` "location" ``of`` (``ident`` "T")] ``)``
+                                    [``public``;``abstract``]
+                                    ``{`` [] ``}`` :> MemberDeclarationSyntax
+        let dumpRegisterMethod = ``method`` "void" "DoDumpRegister" ``<<`` ["T"] ``>>``
+                                    ``(`` [``param`` "location" ``of`` (``ident`` "T");
+                                           ``param`` "qbs" ``of`` qubitArrayType] ``)``
+                                    [``public``;``abstract``]
+                                    ``{`` [] ``}`` :> MemberDeclarationSyntax
+        let nameProperty = ``property-arrow_get`` "string" "Name" [``public``;``override``] 
+                                ``get`` (``=>`` (literal targetClass))
+        let start = if generateDump
+                    then seq { nameProperty :> MemberDeclarationSyntax; 
+                               dumpMachineMethod; dumpRegisterMethod }
+                    else seq { nameProperty :> MemberDeclarationSyntax }
+        if generateQubitMgmt then
+            seq {
+                    start
+                    buildAllocMethods "Allocate"
+                    buildReleaseMethods "Return"
+                    buildReleaseMethods "Release"
+                    buildAllocMethods "Borrow" 
+                } 
+                |> Seq.concat
+        else
+            start
+
+    let buildEpilogue targetClass generateQubitMgmt generateDump =
+        let buildSimField () =
+            ``field`` targetClass "simulator" [``private``] None
+        let buildConstructor name =
+            ``constructor`` name ``(`` [ ("m", ``type`` targetClass) ] ``)`` ``:`` ["m"]
+                    [``public``] 
+                    ``(`` [ (``ident`` "simulator" <-- ``ident`` "m") |> statement ] ``)``
+        let buildAllocClass name =
```

Comment:
nit: I'd split the larger inner helper functions out into separate private functions so that it is easier to read through the actual buildEpilogue function.